### ISSUE TITLE
"Bind Scope" does not apply to infix notations

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#19676]
Consider this example:
```coq
From Coq Require Import String.
From Coq Require Export ZArith.

(** ** Expressions / Terms. *)
Inductive expr :=
  (* Base lambda calculus *)
  | Var (x : string)
  | Lam (x : string) (e : expr)
  | App (e1 e2 : expr)
  (* Base types and their operations *)
  | LitInt (n: Z)
  | Plus (e1 e2 : expr).

Inductive step : expr -> expr -> Prop :=
  | StepPlusR e1 e2 e2' :
    step e2 e2' ->
    step (Plus e1 e2) (Plus e1 e2').
Infix "≻" := step (at level 90).

Declare Scope expr_scope.
Delimit Scope expr_scope with E.
Bind Scope expr_scope with expr.

Notation "e1 + e2" := (Plus e1 e2) : expr_scope.

Definition ZERO := LitInt 0.
Check step (ZERO + ZERO) ZERO. (* this works *)
Check (ZERO + ZERO) ≻ ZERO. (* this errors since it interprets the `+` the wrong way *)
```
We have bound the expr_scope to expr, so as expected the arguments of `step` are parsed in that scope.

However, when I use the infix notation for `step`, then this is not the case. That is quite non-intuitive as it's still the same underlying term. It's also annoying as it means we have to re-declare the notation like so:
```
Notation "e1 ≻ e2" := (step e1%E e2%E) (at level 90, e1, e2 at level ???).
(* I don't even know which level here would be equivalent to the `Infix`. *)
```
which is a lot more verbose.

**Comment by SkySkimmer** [comment]
Isn't it rather that Bind Scope is not retroactive? If you do Bind Scope then the Infix it works.

**Comment by RalfJung** [comment]
*Oh*. I didn't realize that. Since `Bind Scope` is retroactive for definitions it didn't occur to me that it could possibly *not* be retroactive for notations. Thanks for pointing that out!

This does not seem to be currently documented?

Consider this bug report a feature request then to make the order of `Notation` and `Bind Scope` not matter. :)

**Comment by RalfJung** [comment]
Cc @robbertkrebbers -- I think this explains why we had to defensively add `%scope` to tons of our notations.

**Comment by SkySkimmer** [comment]
TBH I'm surprised it's retroactive for references, I guess it's from this code https://github.com/coq/coq/blob/8b24f0f303dfa21051944a4174be0b4bde5fc153/interp/notation.ml#L2080

**Comment by SkySkimmer** [comment]
I'm not sure how you would handle inconsistent scopes if dynamically re inferring scopes for notations. ie in
~~~coq

Axioms T1 T2 : Type.

Axiom foo : T1 -> T2 -> Prop.

Notation "'bar' x" := (foo x x) (at level 42).

Bind Scope type_scope with T1.
~~~
if the notation and bind scope are swapped we get
~~~
Warning: Argument x was previously inferred to be in scope type_scope but is here used in
the empty scope stack. Scope type_scope will be used at parsing time unless you override it by
annotating the argument with an explicit scope of choice. [inconsistent-scopes,syntax,default]
~~~
on the notation. If Bind Scope was retroactive what should happen?

**Comment by RalfJung** [comment]
Hm... it could show that warning, or it could only retroactively change notations where there is no such problem (which is the vast majority of notations, I think).

**Comment by SkySkimmer** [comment]
The update is lazy so a warning would be at the first use of the notation after Bind Scope, not on Bind Scope
I guess that may be reasonable UX

**Comment by robbertkrebbers** [comment]
> Cc @robbertkrebbers -- I think this explains why we had to defensively add %scope to tons of our notations.

I don't think so. We mostly use %scope when we have definitions involving functions, e.g., https://gitlab.mpi-sws.org/iris/iris/-/blob/master/iris/bi/interface.v?ref_type=heads#L343 But I sure could be overlooking other uses.
